Add MooseServer definitions, formatting, and VS Code updates

brandonlangley opened the pull request:
This includes the following updates with the MOOSE language server.

 - full document formatting capability

 - goto-definition capability including
   - look up of definitions from the current or !included input files
   - jump to source file where the object was registered for types
   - send empty response if no file exists at path for source code

 - fix-ups from feedback via VS Code
   - initialize with the current set of supported server capabilities
   - strip arbitrary URI schemes fixing diagnostic line and column
   - remove completion description lead spaces and (REQUIRED)
   - set SymbolKind value on document symbols for outline icons
   - set CompletionItemKind values used by completion list icons

Here is NEAMS Workbench demonstrating definitions and formatting.

https://github.com/idaholab/moose/assets/21372118/f127cf3b-f7a1-4927-9c38-5e5ab8943a0a

refs #22766

brandonlangley committed:
Implement full go-to-definitions capabilities for language server

refs #22766

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1724715/) on d3fa898 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a5148c</th>
      <th colspan=4>#25305 d3fa89</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.65%</td>
      <td>85.65%</td>
      <td>-0.00%</td>
      <td>85.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90166</td>
      <td>90218</td>
      <td>+52</td>
      <td>54</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15109</td>
      <td>15119</td>
      <td>+10</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25305/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25305/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25305/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.71% is less than the suggested 90.0%



This comment will be updated on new commits.

loganharbour commented:
Remove the draft status when you're ready for a review and I'll take a look.

brandonlangley committed:
Implement entire document format capabilities for language server

refs #22766

brandonlangley commented:
@loganharbour -

This adds both go-to-definitions and input-formatting capabilities to the MooseServer.

I had to make a change to WASP for document formatting which is still awaiting merge.

After that's merged, I'll update the WASP submodule here and remove the draft status.

brandonlangley commented:
@loganharbour -

This is ready for review now and is a much smaller change-set than the previous PRs split into these four commits:

1. Goto-Definitions capability implementation and unit testing

2. Input-Formatting capability implementation and unit testing

3. Needed update to WASP submodule and Conda YAML files

4. Bump in versioner_hashes.yaml as a result of WASP update

Here's a short video showing the Goto-Definitions then Input-Formatting features in Workbench using this branch.

https://github.com/idaholab/moose/assets/21372118/960724e6-436c-4e8f-bbeb-cbc1321318e4

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1755103/) on 6c074dd wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 82.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 64.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1755103/) on 6c074dd wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 82.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 34.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 64.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

dschwen commented:
We indent .i files with two spaces. Is that configurable? It should be the default to avoid creating large diffs when editing files that have been formatted with `hit format`.

brandonlangley commented:
@dschwen -

> We indent .i files with two spaces. Is that configurable? It should be the default to avoid creating large diffs when editing files that have been formatted with `hit format`.

Right, the `hit::Formatter::indent_string` is defaulted to `"  "` which I think is a good formatting indention default.

The changes here don't modify that default or make any changes at all to the `hit::Formatter`.

A `hit::Formatter` is just used by the `MooseServer` when a client requests formatting over the language server protocol.

And the language server protocol document formatting request actually [requires](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#formattingOptions) the `tabSize` parameter to be specified.

So any time a client / IDE requests document formatting, they will also always have to supply this `tabSize`  parameter.

The `MooseServer` logic that handles the formatting request simply uses the integer value of that required parameter:

```cpp
  hit::Formatter formatter("cfg", "[format]indent_string='" + std::string(tab_size, ' ') + "'[]");
  std::string document_formatted = formatter.format(document_path, document_text);
```

It seems like an odd choice in the protocol that clients are forced to pick a formatting indention size as mentioned here:

https://github.com/microsoft/language-server-protocol/issues/1166#issuecomment-957325352

It would be nice if it allowed a client to defer that decision to the language server that is actually doing the formatting.

In the `MooseServerTest`, I set `tab_size = 4` since the default indention size in the `hit::Formatter` is two spaces.

And I wanted to test formatting with a value different from two to ensure the requested size is [actually being applied](https://github.com/idaholab/moose/pull/25305/commits/da6109a6fa8eeed9469370e6a0c5d0bd3a7177ca#diff-e861ff6bdefa1faecc5dfe43d534b0008bfeabd7d6895ac6dad76c4cd088c0faR1570-R1628).

dschwen commented:
Can you show that this works with actions? E.g. variables generated by actions. It would be nice to retain the current plugin's behavior of jumping to source code for `type` definitions.

brandonlangley commented:
@dschwen 

> Can you show that this works with actions? E.g. variables generated by actions. It would be nice to retain the current plugin's behavior of jumping to source code for `type` definitions.

Right now the Go-to-definition implementation provides definitions within the input file for referential lookup values, e.g.,

```
01    [Variables]
02      [u]
03        family = MONOMIAL
04        order = CONSTANT
05      []
06    []
07    [BCs]
08      [all]
09        type = VacuumBC
10        boundary = 'left right'
11        variable = u
12      []
13    []
```

with the cursor on the value `u` on line 11, `Go to definition` moves the cursor to select the definition of `[u]` on line 2.

So this capability does not currently provide any response locations when used on the value of a  block `type` parameter.

I like this idea though and I see how the `getLineInfo()` calls are being used to add `file_info` to the JSON syntax tree.

Since `type` parameter values are unused in this feature now, I can add this without conflicting with any current behavior.

The one question I have is with unit testing a MooseServer response since locations consist of file-path and line-number.

Is it okay to test a specific location since the line-number will be invalid if even one line gets added that source code file?

brandonlangley committed:
Fix symbol tree to only contain current document without includes

refs #22766

brandonlangley committed:
Remove any arbitrary document URI scheme rather than only wasplsp

refs #22766

brandonlangley committed:
Add set of currently supported server capabilities for initialize

refs #22766

brandonlangley committed:
Set document symbol kinds which client may use for icons in trees

refs #22766

brandonlangley committed:
Set values to same document symbol kind to allow client filtering

refs #22766

brandonlangley committed:
Set completion item kinds which client may use for icons in lists

refs #22766

brandonlangley committed:
Remove lead spacing and required tag from completion descriptions

refs #22766

brandonlangley committed:
Respond to go-to-definitions with target file instead of base URI

refs #22766

brandonlangley committed:
Use source code location for go-to-definitions of type parameters

refs #22766

brandonlangley commented:
@dschwen -

This is finished now and ready for your review plus any further testing in VS Code that you would like - it includes:

 - full document formatting capability

 - goto-definition capability including
   - look up of definitions from the current or !included input files
   - jump to source file where the object was registered for types
   - send empty response if no file exists at path for source code

 - fix-ups via your VS Code feedback
   - initialize with the current set of supported server capabilities
   - strip arbitrary URI schemes fixing diagnostic line and column
   - remove completion description lead spaces and (REQUIRED)
   - set SymbolKind value on document symbols for outline icons
   - set CompletionItemKind values used by completion list icons

I think more complex pieces from https://github.com/idaholab/moose/issues/25787 like action generated names or block restrictions can wait on a later PR.

Regarding SymbolKind outline icons and CompletionItemKind list icons, I found this image of the full VS Code set:

![CompletionItemKind-SymbolKind-VSCode-Icons](https://github.com/idaholab/moose/assets/21372118/5befe9e0-ba23-485a-b369-2673c7f879e8)

brandonlangley commented:
@dschwen -

FYI, I'm planning to wait until after your review and testing to rebase this branch and fix the merge conflicts if that's okay.

brandonlangley commented:
@loganharbour -

I just rebased this fixing the merge conflicts and dropped the `wasp`, `conda`, and `versioner_hashes` commits.

This is finished and ready to merge at your convenience.

loganharbour commented on framework/src/base/MooseServer.C:
```diff
@@ -721,13 +743,177 @@ MooseServer::getEnumsAndDocs(MooseEnumType & moose_enum_param,
+  std::string param_name = request_context.has_parent() ? request_context.parent().name() : "";
+  std::string val_string = request_context.last_as_string();
+
+  // add source code location if type parameter with registered object name
```
Does this link object in input -> source file?

brandonlangley replied:
@loganharbour -

Yes, this makes `type=ObjectName` work like goto-definition in @dschwen's VSCode plugin as he [requested](https://github.com/idaholab/moose/pull/25305#issuecomment-1749286453).

It returns the path to the `ObjectName.C` source code file where the object was registered for the GUI to use.

I talked to him previously about what to do when the source file does not exist as in executable-only installs.

And it was decided the best thing to do in that situation is send an empty response, so this check does that:

```cpp
    // return without any definition locations added if file cannot be read
    if (!file_line_info.isValid() ||
        !MooseUtils::checkFileReadable(file_line_info.file(), false, false, false))
```

This type --> source scenario is unit tested in the `DefinitionObjectTypeSource` case of `MooseServerTest`.

loganharbour reviewed (approved):
Looks good to me. I think that a @dschwen once-through is a good idea, though.

brandonlangley committed:
Trim trailing spaces from diagnostic messages in server unit test

refs #22766

brandonlangley commented:
@dschwen -

I just rebased this and fixed the `MooseServerTest` merge conflict so this is ready for you to look at again and merge.
